### PR TITLE
Do not throw an exception if the process finished quickly but without any error.

### PR DESCRIPTION
**Pull request by przemekwitek**
Relates https://github.com/elastic/elasticsearch/issues/45787

### CODE REVIEW

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/2

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/MemoryUsageEstimationProcessManager.java [hunk 1]
```diff
@@ -83,9 +84,14 @@ private MemoryUsageEstimationResult runJob(String jobId,
                 onProcessCrash(jobId, processHolder));
         processHolder.process = process;
         if (process.isProcessAlive() == false) {
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/MemoryUsageEstimationProcessManager.java]: It seems to me that the problem that led to this change reveals that this condition is inappropriate.

Sometimes the process has exited before we read its result (from the named pipe which _is_ still alive) and sometimes it is still running when we try to read its result.

This is all down to OS thread and process scheduling, so which of these scenarios occurs shouldn't make a difference to the outcome of this method.  Only the output/errors of the C++ process should make a difference.

So it seems to me that the `String processError = process.readError();` and the lines that handle the value of `processError` should be in the `} catch (Exception e) {` block after trying to read the result.

In other words:

1. Try to read the result
2. If that finds 1 result it returns and we return that to our caller
3. If that finds 0 or 2+ results then it throws an exception, and at that point we check whether errors from the C++ reveal something more useful than simply saying 0 or >1 result.

Comment (przemekwitek) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/MemoryUsageEstimationProcessManager.java]: Done.

Additionally, I've just realized that `onProcessCrash` handler will never be executed as I  added check which prevents that in one of my older PRs. This allowed to further simplify the code.

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
I changed the label from `>bug` to `>non-issue` as this functionality didn't exist prior to 7.4.0 so this fix doesn't need to be release noted (providing it's merged in time for 7.4.0).
